features/support/steps.js

```diff
@@ -25,8 +25,24 @@ Before(function hook() {
 });
 
 After(async function hook() {
-  fs.remove(this.dir);
-  await util.promisify(kill)(this.proc.pid);
+  // kill all running processes
+  const pids = [];
+  try {
+    pids.push(...await pidtree(this.proc.pid));
+  } catch (error) {
+    // the process doesn't exist anymore
+  }
+  pids.forEach((pid) => {
+    try {
+      process.kill(pid, 'SIGKILL');
```

Comment:
I'd like to keep this as minimal as possible - it is an after-test tear down cleanup. I know SIGKILL is brutal, but I'd say tests tear down is allowed to use brutal means to cleanup after tests on two lines of code. There are three situations atm when killing can happen:

1. Dredd malfunction. Dredd isn't able to kill it's children and they get stuck as running processes. I wonder whether this would even track the children down in case Dredd, the parent process, is already finished 🤔 But it's the best I have right now... in any way, that would be malfunction, and those can cause other malfunctions in chain reaction (in this case malfunction of tests cleanup), that's kind of expected.
2. Dredd itself gets stuck in limbo because of some unexpected malfunction. This should definitely kill it and all its children with no mercy and I think that's fine.
3. There are some tests running the hooks handler as a standalone process, sending TCP messages its way and inspecting what it returns. Then they rely on this cleanup to tear down the process (in user environment one would <kbd>Ctrl+C</kbd> to `SIGINT` it).

The test suite template could add to the contract that the hooks handler process should be able to handle SIGTERM, but I'd see that as a separate feature, an addition to what this test suite imposes on hooks handlers implementation. I filed https://github.com/apiaryio/dredd-hooks-template/issues/67 to discuss that. Also, now it could make some of the hooks handlers not passing the test suite.

However, I think we could use your comment (and perhaps the additional code you sent me over Slack?) as a suggestion to improve how Dredd's subprocess management works. Currently it lives here: https://github.com/apiaryio/dredd/blob/72220703a644234436c53de42849ab7fa045653a/lib/childProcess.js Sometimes it feels flaky in the CI and it still has bugs, even though it's much more reliable than it was previously (before I revamped it and put into `childProcess.js`). Because there, in Dredd, that's IMHO the place where care should be taken and where exactly what you write happens - nicely trying SIGTERM first, waiting, then killing with SIGKILL, etc.